packages/plugins/response-cache/src/redis-cache.ts

```diff
@@ -0,0 +1,115 @@
+   * By default the typename is concatenated with the id e.g. `User:1`
+   */
+  buildRedisEntityId?: BuildRedisEntityId;
+  /**
+   * Customize how the cache key that stores the operations associated with the response is built.
+   * By default `operations` is concatenated with the responseId e.g. `operations:arZm3tCKgGmpu+a5slrpSH9vjSQ=`
+   */
+  buildRedisResponseOpsKey?: BuildRedisResponseOpsKey;
+};
+
+export const createRedisCache = (params: RedisCacheParameter): Cache => {
+  const store = params.redis;
+
+  const buildRedisEntityId = params?.buildRedisEntityId ?? defaultBuildRedisEntityId;
+  const buildRedisResponseOpsKey = params?.buildRedisResponseOpsKey ?? defaultBuildRedisResponseOpsKey;
+
+  function purgeEntity(entity: string) {
+    // find the responseIds for the entity
+    store.smembers(entity).then(function (responseIds) {
+      // and purge each response since they contained the entity data
+      store.del(
+        responseIds.map(responseId => {
+          return buildRedisResponseOpsKey(responseId);
+        })
+      );
+      store.del(responseIds);
+    });
+
+    // if purging an entity like Comment, then also purge Comment:1, Comment:2, etc
+    if (!entity.includes(':')) {
+      store.keys(`${entity}:*`).then(function (entityKeys) {
+        for (const entityKey of entityKeys) {
+          // and purge any reponses in each of those entity keys
+          store.smembers(entityKey).then(function (responseIds) {
+            // if purging an entity check for associated operations containing that entity
+            store.del(
+              responseIds.map(responseId => {
+                return buildRedisResponseOpsKey(responseId);
+              })
+            );
+            // and purge each response since they contained the entity data
+            store.del(responseIds);
+          });
+        }
+
+        // then purge the entityKeys like Comment:1, Comment:2 etc
+        store.del(entityKeys);
+      });
+    }
+
+    // and then purge the entity itself
+    store.del(entity);
+  }
+
+  return {
+    set(responseId, result, collectedEntities, ttl) {
+      if (ttl === Infinity) {
+        store.set(responseId, JSON.stringify(result));
+      } else {
+        // set the ttl in milliseconds
+        store.set(responseId, JSON.stringify(result), 'PX', ttl);
+      }
+
+      const responseKey = buildRedisResponseOpsKey(responseId);
+
+      for (const { typename, id } of collectedEntities) {
+        // Adds a key for the typename => response
+        store.sadd(typename, responseId);
+        // Adds a key for the operation => typename
+        store.sadd(responseKey, typename);
+
+        if (id) {
+          const entityId = buildRedisEntityId(typename, id);
+          // Adds a key for the typename:id => response
+          store.sadd(entityId, responseId);
+          // Adds a key for the operation => typename:id
+          store.sadd(responseKey, entityId);
+        }
+      }
```

Comment:
```suggestion
      const pipeline = store.pipeline();
      if (ttl === Infinity) {
        pipeline.set(responseId, JSON.stringify(result));
      } else {
        // set the ttl in milliseconds
        pipeline.set(responseId, JSON.stringify(result), 'PX', ttl);
      }

      const responseKey = buildRedisResponseOpsKey(responseId);

      for (const { typename, id } of collectedEntities) {
        // Adds a key for the typename => response
        pipeline.sadd(typename, responseId);
        // Adds a key for the operation => typename
        pipeline.sadd(responseKey, typename);

        if (id) {
          const entityId = buildRedisEntityId(typename, id);
          // Adds a key for the typename:id => response
          pipeline.sadd(entityId, responseId);
          // Adds a key for the operation => typename:id
          pipeline.sadd(responseKey, entityId);
        }
      }
      
      pipeline.exec()
```

Reply:
diff looks awkward... I used`store.pipeline()` replaced `store.sadd` with `pipeline.sadd` and added `pipeline.exec()` at the end.

Reply:
I also had thought of ... well a transaction to surround the adds ... but pipeline is better. Will implement.